packages/grid/x-data-grid/src/hooks/core/useGridApiInitialization.ts

```diff
@@ -11,14 +16,53 @@ const isSyntheticEvent = (event: any): event is React.SyntheticEvent => {
 
 let globalId = 0;
 
-export function useGridApiInitialization<Api extends GridApiCommon>(
+const wrapPublicApi = <PrivateApi extends GridPrivateApiCommon, PublicApi extends GridApiCommon>(
+  publicApi: PublicApi,
+) => {
+  type PrivateOnlyApi = GridPrivateOnlyApiCommon<PublicApi, PrivateApi>;
+  const privateOnlyApi = {} as PrivateOnlyApi;
+
+  privateOnlyApi.getPublicApi = () => publicApi;
+
+  privateOnlyApi.register = (visibility, methods) => {
+    Object.keys(methods).forEach((methodName) => {
+      if (visibility === 'public') {
+        if (!publicApi.hasOwnProperty(methodName)) {
+          publicApi[methodName as keyof PublicApi] = (methods as any)[methodName];
+        }
+      } else if (!privateOnlyApi.hasOwnProperty(methodName)) {
+        privateOnlyApi[methodName as keyof PrivateOnlyApi] = (methods as any)[methodName];
+      }
+    });
+  };
+
+  const handler: ProxyHandler<GridApiCommon> = {
+    get: (obj, prop) => {
+      if (prop in obj) {
+        return obj[prop as keyof typeof obj];
+      }
+      return privateOnlyApi[prop as keyof PrivateOnlyApi];
+    },
+    set: (obj, prop, value) => {
+      obj[prop as keyof typeof obj] = value;
+      return true;
+    },
+  };
+
+  return new Proxy(publicApi, handler) as PrivateApi;
+};
+
+export function useGridApiInitialization<
+  PrivateApi extends GridPrivateApiCommon,
+  Api extends GridApiCommon,
+>(
   inputApiRef: React.MutableRefObject<Api> | undefined,
   props: Pick<DataGridProcessedProps, 'signature'>,
-): React.MutableRefObject<Api> {
-  const apiRef = React.useRef() as React.MutableRefObject<Api>;
+): React.MutableRefObject<PrivateApi> {
+  const publicApiRef = React.useRef() as React.MutableRefObject<Api>;
 
-  if (!apiRef.current) {
-    apiRef.current = {
+  if (!publicApiRef.current) {
+    publicApiRef.current = {
       unstable_eventManager: new EventManager(),
       unstable_caches: {} as Api['unstable_caches'],
       state: {} as Api['state'],
```

Comment:
Yes, I would do this in the follow up PR (except for `state` - it will remain public)

Reply:
> The only use for state is to call simple selectors.

Agree, that's the only reason for it to be public

> We could rename it to unstable_state and give a wrapper function that calls the selector passing the right arguments.

Yes, we could, but why?
This would be a breaking change for anyone using selectors and `useGridSelector` won't work anymore with public `apiRef`.
Also, in https://mui.com/x/react-data-grid/state/#access-the-state we mention that `apiRef.current.state` is not considered a public API.